<h3 style="text-align: center;">Бинарная и мультиклассовая классификация на крауде</h3>

Домашнее задание №4 по дисциплине "Сбор данных с помощью краудсорсинга"  

**Студент:** &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Погребная Александра  
**Группа:**  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  794, кафедра АД МФТИ

## Задание <a class="anchor" id="first-bullet"></a>

<img src="https://www.dropbox.com/s/qtewugawp3wdb2r/2020-10-22_03-06-01.png?raw=1" width="500px" alt="">

## Отчёт

*Если захотите воспроизвести результаты у себя, то можете скачать обработанные данные здесь:

In [1]:
# !wget -O hw4.zip https://www.dropbox.com/s/gjxr7lw644l2mxr/hw4.zip?dl=0
# !unzip hw4.zip

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from collections import Counter

**1. Подготовка проектов**

**full** - первый лист таблицы из условия, но без правильных ответов, это выборка, которую нужно классифицировать.  
**exam** - второй лист таблицы.

Для мультиклассовой классификации создала проект с обучением, экзаменом и основным пулом: в качестве данных взяла просто full и exam с GOLDEN (и HINT) без ответов. Чтобы подготовить данные, использовала Excel.  
Для бинарной классификации решила сделать один проект, в котором будет по три пула на каждый дорожный знак, т.е.:
+ 5 обучающих пулов: по 10 обучающих из exam в задании
+ 5 экзаменационных: по 20 контрольных из exam в задании - здесь произошел fail, но об этом позже
+ 5 основных: по 40 тестовых из full и по 4 контрольных из exam в задании

Код ниже генерирует данные для этих пулов

In [3]:
# exam = pd.read_csv("./exam.tsv", sep='\t', header=0, encoding="UTF-8")
# full = pd.read_csv("./full.tsv", sep='\t', header=0, encoding="UTF-8")
# print("train/exam columns: {}".format(exam.columns))
# print("test columns: {}".format(full.columns))
# exam

In [4]:
signs = {
    "pedestrian": "Пешеходный переход",
    "bus_stop":   "Остановка общественного транспорта",
    "no_stop":    "Остановка запрещена",
    "main_road":  "Главная дорога",
    "give_way":   "Уступи дорогу"
}

Код для создания данных для бинарной классификации:

In [5]:
# for sign in signs.keys():
#     # tsv для итогового пула
#     sign_full_label = [signs[sign]] * len(full['INPUT:image'])
#     sign_full_data = full
#     sign_full_data["INPUT:label"] = sign_full_label
#     sign_full_data.to_csv(sign+"_full.tsv", index=False, sep='\t', encoding="UTF-8")

#     # tsv для экзамена и обучения
#     hints = []
#     sign_golden = []
#     for ans in exam["ans"]:
#         if ans == sign:
#             hints.append("Да")
#             sign_golden.append("true")
#         else:
#             hints.append("Нет")
#             sign_golden.append("false")
#     sign_exam = exam.drop(columns=["ans"])

#     sign_exam_label = [signs[sign]] * len(exam)
#     sign_exam["INPUT:label"] = sign_exam_label

#     # экзамен
#     sign_exam["GOLDEN:result"] = sign_golden
#     sign_exam.to_csv(sign+"_exam.tsv", index=False, sep='\t', encoding="UTF-8")

#     # обучение
#     sign_exam["HINT:text"] = hints
#     sign_exam.to_csv(sign+"_train.tsv", index=False, sep='\t', encoding="UTF-8")

Получившиеся проекты для классификации:
+ **мультиклассовой:** https://toloka.yandex.ru/requester/project/42164
+ **бинарной:** https://toloka.yandex.ru/requester/project/42359

Чек-лист:  
✅ создала проекты  
✅ написала инструкции  
✅ создала пулы  
✅ создала ханипоты  
✅ настроила контроль качества  
✅ запустила проекты  
✅ получила результаты  

**2. Обработка результатов**

**2.1. Стоимость**

Расходы мультиклассового проекта:     0.2 (+ 0.1) = **0.3\\$**  
Стоимость одной картинки:   0.3 / 250 = **0.12 цента**

Расходы бинарного проекта: (0.14 (+ 0.07))\*4 + 0.15 (+ 0.07) = **1.06\\$**  
Стоимость одной картинки: 1.06 / 250 = **0.424 цента**

Вывод: бинарная классификация получилась примерно в 3.5 раза дороже.

**2.2. Отбор исполнителей**

Мною во всех экзаменационных пулах была допущена ошибка - я не добавила следующий блок в контроль качества:

<img src="https://i.ibb.co/XZsLG6Q/bug.png" alt="bug" border="0"></a> - его не было в инструкции к заданию, но добавлять его осмысленно.

Я хотела, чтобы во всех экзаменах было по 20 заданий, а допуск к основному заданию открывался при 85% правильных ответов - это 17/20. Однако теперь толокеры сдавали от 1 до 3 страниц заданий, т.е. от 20 до 60, причем навык назначался только по результатам последней сданной страницы. Заметила это только при обработке результатов.

Посмотрим, какая доля исполнителей успешно проходила экзамены.  
Раздел "Статистика" в Толоке в мультиклассовом экзамене показывает качество выше 0.9:

<img src="https://i.ibb.co/8jhWwxC/image.png" alt="" border="0"></a>

В бинарных пулах по-разному, например, в экзамене знака "Остановка запрещена" качество выше 0.98:

<img src="https://i.ibb.co/GMfCsnS/image.png" border="0"></a>

А в экзамене "Уступи дорогу" есть и 0.72:

<img src="https://i.ibb.co/v40WB7V/image.png" alt="" border="0"></a>

Но это усредненные значения. Посмотрим подробнее, для этого используем результаты всех экзаменов.  

Проще было сразу скачать данные из раздела "Навыки" в Толоке. В моем случае это даст посмотреть на тех, кто сдал экзамен, но не на тех, кто действительно отвечал на все вопросы хорошо, поскольку у меня "Навык" учитывает лишь последнюю экзаменационную страницу.  
Так что посмотрим на все ответы исполнителей и проверим: возможно кто-то из них случайно хорошо разметил последнюю страницу экзамена, но плохо справился с первыми. Тогда его разметку основного задания не стоит рассматривать.

Сначала бинарная классификация:

In [6]:
exam_binary_files = ["./ass_exam_" + sign + ".tsv" for sign in signs.keys()]

binary_exams = pd.concat((pd.read_csv(f, sep='\t', header=0, encoding="UTF-8") 
                          for f in exam_binary_files))
binary_exams

,INPUT:image,INPUT:label,OUTPUT:result,GOLDEN:result,HINT:text,ASSIGNMENT:worker_id
0,http://sdcimages.s3.yandex.net/nirvana/e127774...,Пешеходный переход,False,False,NaN,70c6056722e434ee9c0b68267181b07c
1,http://sdcimages.s3.yandex.net/nirvana/628c77c...,Пешеходный переход,False,False,NaN,70c6056722e434ee9c0b68267181b07c
2,http://sdcimages.s3.yandex.net/nirvana/28873dc...,Пешеходный переход,False,False,NaN,70c6056722e434ee9c0b68267181b07c
3,http://sdcimages.s3.yandex.net/nirvana/3c267a7...,Пешеходный переход,True,True,NaN,70c6056722e434ee9c0b68267181b07c
4,http://sdcimages.s3.yandex.net/nirvana/8993c14...,Пешеходный переход,False,False,NaN,70c6056722e434ee9c0b68267181b07c
...,...,...,...,...,...,...
1375,http://sdcimages.s3.yandex.net/nirvana/5d4d302...,Уступи дорогу,False,False,NaN,69e918870fcea9f899e616e1845b59c4
1376,http://sdcimages.s3.yandex.net/nirvana/8993c14...,Уступи дорогу,False,False,NaN,69e918870fcea9f899e616e1845b59c4
1377,http://sdcimages.s3.yandex.net/nirvana/3fee0d6...,Уступи дорогу,False,False,NaN,69e918870fcea9f899e616e1845b59c4
1378,http://sdcimages.s3.yandex.net/nirvana/5e8b4f5...,Уступи дорогу,False,False,NaN,69e918870fcea9f899e616e1845b59c4


In [7]:
binary_exams["result"] = (binary_exams["OUTPUT:result"] == binary_exams["GOLDEN:result"])
skills = binary_exams[["result", "INPUT:label", "ASSIGNMENT:worker_id"]].\
                groupby(["INPUT:label", "ASSIGNMENT:worker_id"]).agg(['mean'])
skills

result
                                                     mean
INPUT:label    ASSIGNMENT:worker_id                      
Главная дорога 03c08ada934af11ff8561cc39fe15f7c  1.000000
               08c023f84fe4e4547d1337f389b61bc9  1.000000
               1bcfb516f4d28bac06426a9b2707f772  1.000000
               2e37aab3acb35fb0450ba86bb46a3e13  1.000000
               327471a9a9745898fc1200f83208b763  1.000000
...                                                   ...
Уступи дорогу  d7cf3c255e94a112e71824aca72e48e1  1.000000
               d9b37fda2b4de2f891735bbb0b371582  1.000000
               dd7c7399d42f00609410069f5b4c2351  1.000000
               faa058810c8a490660ab6c0442679995  0.983333
               fb1b6e26cd8574da4dc61681a396f47c  1.000000

[121 rows x 1 columns]

In [8]:
good_res_num = (skills["result"]["mean"] >= 0.85).sum()
res_num = len(skills["result"]["mean"])

print("Доля исполнителей, успешно решавших бинарный экзамен: {:.4f}".format(good_res_num / res_num))

Доля исполнителей, успешно решавших бинарный экзамен: 0.9917


In [9]:
print(res_num - good_res_num)

1


In [10]:
skills[skills["result"]["mean"] < 0.85]

,,result
,,mean
INPUT:label,ASSIGNMENT:worker_id,
Уступи дорогу,1bcfb516f4d28bac06426a9b2707f772,0.583333


In [11]:
bad_user = binary_exams.loc[binary_exams["ASSIGNMENT:worker_id"] == "1bcfb516f4d28bac06426a9b2707f772"]
bad_user = bad_user.loc[bad_user["INPUT:label"] == "Уступи дорогу"]
print("Не сдавший человек сдавал {} заданий, ответил верно на {} из них.".\
      format(len(bad_user), bad_user["result"].sum()))

Не сдавший человек сдавал 60 заданий, ответил верно на 35 из них.


In [12]:
print(res_num)

121


Не прошел всего один исполнитель из 121, как раз тот, из-за которого просело качество в экзамене на "Уступи дорогу". Это не бот, потому что он же сдавал экзамены на пешеходные переходы, главные дороги и остановки на навыки 90, 100 и 100 соответственно:

<img src="https://i.ibb.co/s9nG8K2/2020-10-23-06-11-41.png" border="0"></a>

Видимо просто устал. Навык у него равен 60, поэтому в экзамен он в любом случае не попал. Таким образ, не смотря на то, что проходной навык учитывал лишь последнюю сданную страницу, в основной пул все равно попали только пользователи, успешно прорешавшие $\geq$ 85% заданий.

Теперь разберемся с мультиклассовыми экзаменами:

In [13]:
multi_exams = pd.read_csv("ass_exam_multi.tsv", sep='\t', header=0, encoding="UTF-8") 

multi_exams

,INPUT:image,OUTPUT:result,GOLDEN:result,HINT:text,ASSIGNMENT:worker_id
0,http://sdcimages.s3.yandex.net/nirvana/2835909...,give_way,give_way,NaN,34420b3fc992d699fd0c8dd441c8a909
1,http://sdcimages.s3.yandex.net/nirvana/aae1ec5...,pedestrian,pedestrian,NaN,34420b3fc992d699fd0c8dd441c8a909
2,http://sdcimages.s3.yandex.net/nirvana/e127774...,main_road,main_road,NaN,34420b3fc992d699fd0c8dd441c8a909
3,http://sdcimages.s3.yandex.net/nirvana/a1ce075...,pedestrian,pedestrian,NaN,34420b3fc992d699fd0c8dd441c8a909
4,http://sdcimages.s3.yandex.net/nirvana/bcafe88...,pedestrian,pedestrian,NaN,34420b3fc992d699fd0c8dd441c8a909
...,...,...,...,...,...
3055,http://sdcimages.s3.yandex.net/nirvana/1ae46a8...,pedestrian,pedestrian,NaN,dc0a4cd339525cac8ab9004ddeb83b03
3056,http://sdcimages.s3.yandex.net/nirvana/12c2b63...,no_stop,no_stop,NaN,dc0a4cd339525cac8ab9004ddeb83b03
3057,http://sdcimages.s3.yandex.net/nirvana/3482bca...,main_road,main_road,NaN,dc0a4cd339525cac8ab9004ddeb83b03
3058,http://sdcimages.s3.yandex.net/nirvana/1f87fcf...,give_way,give_way,NaN,dc0a4cd339525cac8ab9004ddeb83b03


In [14]:
multi_exams["result"] = (multi_exams["OUTPUT:result"] == multi_exams["GOLDEN:result"])
multi_skills = multi_exams[["result", "ASSIGNMENT:worker_id"]].\
                    groupby(["ASSIGNMENT:worker_id"]).agg(['mean'])
multi_skills

,result
,mean
ASSIGNMENT:worker_id,
018b5e6241e1cd4628b186f06db9cd2c,1.000000
02b8b4a6ca4b53330207ca942f995470,1.000000
046bcc85c5e7a868b120479416023696,1.000000
0ba5f5086dd7a3e70af43c2662ddcbaf,0.966667
0c3c3c62bb6141cd63a6edf9bbbaa0ee,0.950000
...,...
f139783182cfbdcbf214cd818d249535,1.000000
faa058810c8a490660ab6c0442679995,1.000000


In [15]:
multi_good_res_num = (multi_skills["result"]["mean"] > 0.85).sum()
multi_res_num = len(multi_skills["result"]["mean"])

print("Доля исполнителей, успешно прошедших мультиклассовый экзамен: {:.4f}".\
      format(multi_good_res_num / multi_res_num))

Доля исполнителей, успешно прошедших мультиклассовый экзамен: 1.0000


In [16]:
print(multi_res_num)

68


Итого доля успешно отобранных после экзамена исполнителей:
+ бинарная: **0.9917**
+ мультиклассовая: **1**

Вывод: в бинарной классификации отсеялось больше людей. 

Замечу, что хотя в бинарном проекте было в 5 раз больше пулов, различных толокеров, сдавших экзамен, в нем всего в 2 раза больше. Это может быть связано со временем выкладывания заданий - бинарные пулы я запускала поздно ночью, было мало активных пользователей.

**2.3. Качество разметки**

Теперь посмотрим на accuracy в основных пулах. Верные ответы:

In [17]:
golden_res = pd.read_csv("golden_res.tsv", sep='\t', header=0, encoding="UTF-8").sort_values('INPUT:image')

golden_res

,INPUT:image,GOLDEN:result
19,http://sdcimages.s3.yandex.net/nirvana/0020aec...,pedestrian
126,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,no_stop
139,http://sdcimages.s3.yandex.net/nirvana/01d1f01...,no_stop
100,http://sdcimages.s3.yandex.net/nirvana/020e75c...,no_stop
63,http://sdcimages.s3.yandex.net/nirvana/023c10e...,bus_stop
...,...,...
187,http://sdcimages.s3.yandex.net/nirvana/f90a2fd...,main_road
210,http://sdcimages.s3.yandex.net/nirvana/fa92e0b...,give_way
172,http://sdcimages.s3.yandex.net/nirvana/fc65477...,main_road
155,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,main_road


In [18]:
binary_files = ["./ass_" + sign + ".tsv" for sign in signs.keys()]

binary_res = pd.concat((pd.read_csv(f, sep='\t', header=0, encoding="UTF-8") 
                          for f in binary_files)).sort_values(['INPUT:image', 'INPUT:label'])

assert binary_res["GOLDEN:result"].isna().count() == len(binary_res), "Попали контрольные вопросы"
binary_res = binary_res.drop(columns=["HINT:text", "GOLDEN:result"])
binary_res

,INPUT:image,INPUT:label,OUTPUT:result
446,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False
487,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False
78,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False
172,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False
171,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка общественного транспорта,False
...,...,...,...
427,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Остановка общественного транспорта,False
30,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True
87,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True
393,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Уступи дорогу,False


In [19]:
binary_res = pd.merge(binary_res, golden_res, how="right", on=["INPUT:image"])
binary_res

,INPUT:image,INPUT:label,OUTPUT:result,GOLDEN:result
0,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False,pedestrian
1,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False,pedestrian
2,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False,pedestrian
3,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False,pedestrian
4,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка общественного транспорта,False,pedestrian
...,...,...,...,...
2496,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Остановка общественного транспорта,False,pedestrian
2497,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True,pedestrian
2498,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True,pedestrian
2499,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Уступи дорогу,False,pedestrian


In [20]:
# приводим label и result к одному виду
binary_res["GOLDEN:result"] = binary_res["GOLDEN:result"].apply(lambda x: signs[x])

# если golden и label совпадают, то result должен быть true, иначе - false
# это бинарный golden результат
binary_res["result"] = (binary_res["GOLDEN:result"] == binary_res["INPUT:label"])
binary_res

,INPUT:image,INPUT:label,OUTPUT:result,GOLDEN:result,result
0,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False,Пешеходный переход,False
1,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False,Пешеходный переход,False
2,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False,Пешеходный переход,False
3,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False,Пешеходный переход,False
4,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка общественного транспорта,False,Пешеходный переход,False
...,...,...,...,...,...
2496,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Остановка общественного транспорта,False,Пешеходный переход,False
2497,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True,Пешеходный переход,True
2498,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True,Пешеходный переход,True
2499,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Уступи дорогу,False,Пешеходный переход,False


В задании необходимо сделать итоговый tsv для бинарной классификации:
1. **image** - URL картинки
2. **test_sign** - дорожный знак, которые определяли толокеры
3. **result** - итоговая оценка
4. **overlay** - перекрытие
5. **perc_for_result** - доля оценок из перекрытия за итоговую оценку

In [21]:
binary_final = pd.DataFrame(columns=["image", "test_sign", "result", "overlay", "perc_for_result"])

# URL изображений и проверяемые знаки
binary_groups = binary_res.groupby(["INPUT:image", "INPUT:label"])

images_labes = np.array(list(binary_groups.groups.keys()))
binary_final["image"] = images_labes[:, 0]
binary_final["test_sign"] = images_labes[:, 1]

# перекрытие
binary_final["overlay"] = binary_groups.count()["OUTPUT:result"].to_numpy()

# результат по проверяемому знаку
binary_groups = binary_res[["INPUT:image", "INPUT:label", "OUTPUT:result"]].\
                    groupby(["INPUT:image", "INPUT:label"])

# если больше половины выбранных одинаковы - то это True
def chosen_result(col):
    trues = np.sum(col)
    alls = len(col)
    if trues >= alls // 2:
        return True, trues / alls
    return False, (alls - trues) / alls
    
answers = binary_groups.agg(chosen_result)["OUTPUT:result"].to_numpy()
answers = list(zip(*answers))

binary_final["result"] = answers[0]
binary_final["perc_for_result"] = answers[1]

# pd.merge(binary_final, golden
binary_final

/home/inari/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:22: RuntimeWarning: invalid value encountered in long_scalars


,image,test_sign,result,overlay,perc_for_result
0,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Главная дорога,False,2,1.0
1,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка запрещена,False,2,1.0
2,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Остановка общественного транспорта,False,2,1.0
3,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Пешеходный переход,True,2,1.0
4,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Уступи дорогу,False,2,1.0
...,...,...,...,...,...
1245,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Главная дорога,False,2,1.0
1246,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Остановка запрещена,False,2,1.0
1247,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Остановка общественного транспорта,False,2,1.0
1248,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,Пешеходный переход,True,2,1.0


Сохраню файл с результатами:

In [22]:
binary_final.to_csv("./binary.tsv", sep='\t', encoding="UTF-8")

Посмотрим на accuracy:

In [23]:
golden_results =  binary_res[["INPUT:image", "INPUT:label", "result"]].\
                    groupby(["INPUT:image", "INPUT:label"]).agg(np.mean)["result"].to_numpy()
golden_results

array([False, False, False, ..., False,  True, False])

In [24]:
accuracy = (golden_results == binary_final["result"]).sum() / len(golden_results)

print("Accuracy для бинарной классификации: {}".format(accuracy))

Accuracy для бинарной классификации: 1.0


Круто, вообще без ошибок 👍  
Посмотрим, когда у исполнителей вообще были разногласия в ответах - если были. У меня динамическое перекрытие, т.е. если два исполнителя ответили по-разному, задание давалось третьему:

In [25]:
(binary_final["overlay"] == 3).sum()

1

Всего один неоднозначный результат

In [26]:
binary_final.loc[binary_final["overlay"] == 3]

,image,test_sign,result,overlay,perc_for_result
556,http://sdcimages.s3.yandex.net/nirvana/6dfa257...,Остановка запрещена,True,3,0.666667


Один из исполнителей о вот этом фото ответил, что это не знак "Остановка запрещена":

<img src="http://sdcimages.s3.yandex.net/nirvana/6dfa257f-d6ea-4c9e-b15f-d986d2028e90"> - вероятно просто мисклик

Теперь посмотрим на качество мультиклассовой классификации. Это просто скачанные ответы толокеров:

In [27]:
multi_res = pd.read_csv("ass_multi.tsv", sep='\t', header=0, encoding="UTF-8").sort_values("INPUT:image")
multi_res = multi_res.drop(columns=["GOLDEN:result", "HINT:text"])
multi_res

,INPUT:image,OUTPUT:result
150,http://sdcimages.s3.yandex.net/nirvana/0020aec...,pedestrian
118,http://sdcimages.s3.yandex.net/nirvana/0020aec...,pedestrian
402,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,no_stop
368,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,no_stop
496,http://sdcimages.s3.yandex.net/nirvana/01d1f01...,no_stop
...,...,...
510,http://sdcimages.s3.yandex.net/nirvana/fc65477...,main_road
308,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,main_road
273,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,main_road
26,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,pedestrian


А это результаты разметки, агрегированные по навыку пользователей:

In [28]:
multi_agg = pd.read_csv("agg_multi.tsv", sep='\t', header=0, encoding="UTF-8").sort_values("INPUT:image")
multi_agg = multi_agg.loc[multi_agg["GOLDEN:result"].isnull()]\
                [["INPUT:image", "OUTPUT:result", "CONFIDENCE:result"]]

multi_agg = multi_agg.rename(columns={"OUTPUT:result": "aggregated_result"})

# соединяю с таблицей выше
multi_res = pd.merge(multi_res, multi_agg, how="right", on=["INPUT:image"])
# если агрегированный ответ совпал с ответом пользователя, то True
multi_res["same_output_res"] = (multi_res["aggregated_result"] == multi_res["OUTPUT:result"])
multi_res

,INPUT:image,OUTPUT:result,aggregated_result,CONFIDENCE:result,same_output_res
0,http://sdcimages.s3.yandex.net/nirvana/0020aec...,pedestrian,pedestrian,67.71%,True
1,http://sdcimages.s3.yandex.net/nirvana/0020aec...,pedestrian,pedestrian,67.71%,True
2,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,no_stop,no_stop,67.71%,True
3,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,no_stop,no_stop,67.71%,True
4,http://sdcimages.s3.yandex.net/nirvana/01d1f01...,no_stop,no_stop,67.71%,True
...,...,...,...,...,...
495,http://sdcimages.s3.yandex.net/nirvana/fc65477...,main_road,main_road,67.71%,True
496,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,main_road,main_road,67.71%,True
497,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,main_road,main_road,67.71%,True
498,http://sdcimages.s3.yandex.net/nirvana/ffb1ff5...,pedestrian,pedestrian,67.71%,True


В задании необходимо сделать итоговый tsv для мультиклассовой классификации:
1. **image** - URL картинки
2. **result** - итоговая оценка
3. **overlay** - перекрытие
4. **perc_for_result** - доля оценок из перекрытия за итоговую оценку

In [29]:
multi_final = pd.DataFrame(columns=["image", "result", "overlay", "perc_for_result"])

# URL изображений
multi_groups = multi_res.groupby(["INPUT:image"])
multi_final["image"] = np.array(list(multi_groups.groups.keys()))

# # перекрытие
multi_final["overlay"] = multi_groups.count()["OUTPUT:result"].to_numpy()

# результат, агрегированный по навыкам
multi_groups = multi_res[["INPUT:image", "aggregated_result"]].\
                    groupby(["INPUT:image"])


output_results = list(map(lambda x: signs[x[0]], multi_groups.first().to_numpy()))
multi_final["result"] = output_results

# доля оценок за итоговую оценку
multi_groups = multi_res[["INPUT:image", "same_output_res"]].\
                    groupby(["INPUT:image"])

def part_same_results(col):
    trues = np.sum(col)
    alls = len(col)
    return trues / alls
    
part = multi_groups.apply(part_same_results)["same_output_res"].to_numpy()
multi_final["perc_for_result"] = part

multi_final

,image,result,overlay,perc_for_result
0,http://sdcimages.s3.yandex.net/nirvana/0020aec...,Пешеходный переход,2,1.0
1,http://sdcimages.s3.yandex.net/nirvana/01a2e62...,Остановка запрещена,2,1.0
2,http://sdcimages.s3.yandex.net/nirvana/01d1f01...,Остановка запрещена,2,1.0
3,http://sdcimages.s3.yandex.net/nirvana/020e75c...,Остановка запрещена,2,1.0
4,http://sdcimages.s3.yandex.net/nirvana/023c10e...,Остановка общественного транспорта,2,1.0
...,...,...,...,...
245,http://sdcimages.s3.yandex.net/nirvana/f90a2fd...,Главная дорога,2,1.0
246,http://sdcimages.s3.yandex.net/nirvana/fa92e0b...,Уступи дорогу,2,1.0
247,http://sdcimages.s3.yandex.net/nirvana/fc65477...,Главная дорога,2,1.0
248,http://sdcimages.s3.yandex.net/nirvana/ff1aab5...,Главная дорога,2,1.0


Сохраню файл с результатами:

In [30]:
multi_final.to_csv("./multi.tsv", sep='\t', encoding="UTF-8")

Посмотрим на accuracy:

In [31]:
golden_results = list(map(lambda x: signs[x], golden_res["GOLDEN:result"]))
golden_results[:10]

['Пешеходный переход',
 'Остановка запрещена',
 'Остановка запрещена',
 'Остановка запрещена',
 'Остановка общественного транспорта',
 'Главная дорога',
 'Остановка общественного транспорта',
 'Уступи дорогу',
 'Пешеходный переход',
 'Остановка общественного транспорта']

In [32]:
accuracy = (golden_results == multi_final["result"]).sum() / len(golden_results)

print("Accuracy для мультиклассовой классификации: {}".format(accuracy))

Accuracy для мультиклассовой классификации: 1.0


Круто, здесь тоже без ошибок 👍  
В этом пуле было настроено такое же динамическое перекрытие, как в бинарных. Посмотрим, были ли здесь разные ответы:

In [33]:
(multi_final["overlay"] == 3).sum()

0

Вау, без разногласий!

Вывод: в обоих проектах итоговое 100% качество разметки.

**2.4. Скорость разметки**

Рассмотрим затраченое время на основной пул.

Мультиклассовая классификация:
- среднее время выполнения страницы заданий: 48 сек
- в среднем на одно фото: 48 / 26 = **1.85 сек**
- общее время выполнения: 11 мин 19 сек
- в общем на одно фото: (11 * 60 + 19) / 20 / 26 = **1.31 сек**

---

В бинарной классификации можно считать, что задания в пулах одинаковы, и пулы запускаются параллельно, поэтому логично брать максимальное время из всех.

Бинарная классификация:
- среднее время выполнения страницы заданий: max(55, 77, 59, 57, 96) = 96 сек
- в среднем на одно фото: 96 / 44 = **2.18 сек**
- общее время выполнения: max(12 мин 3 сек, 14 мин 24 сек, 12 мин 21 сек, 7 мин 20 сек, 11 мин 8 сек) = 14 мин 24 сек
- в общем время на одно фото с учетом простоя: (11 * 60 + 19) / 14 / 44 = **1.4 сек**

Предварительный вывод: мультиклассовая классификация выигрывает по скорости.

**Summary**

Бинарная классификация проще многоклассовой. Я ждала, что она:
+ проиграет по стоимости из-за 5 пулов вместо одного;
+ меньше зареджектит исполнителей на экзамене, потому что задание проще;
+ даст лучшее качество и скорость по той же причине.

На практике многоклассовая классификация показала лучшие результаты по всем пунктам. 

Насчет **отбора исполнителей**: думаю, можно считать исполнителя, не сдавшего бинарный экзамен, выбросом. В таком случае оба проекта пропустили 100% людей, сдававших экзамен.

Думаю, **качество** одинаково отличное, т.к. задание разметки дорожных знаков само по себе простое: как бинарное, так и мультиклассовое. Кроме того, пять классов нормально воспринимаются исполнителями - это относительно небольшое количество.

Насчет **стоимости**. Пулов в 5 раз больше, но проигрыш бинарной классификации всего в 3.5 раза - за счет бОльшего количества заданий на странице. Качество от этого не просело, хотя думаю, был один мисклик. Можно было еще уменьшить эту разницу в стоимости, если бы я:
1. сделала 4 пула вместо 5, а последний знак определяла по остаточному признаку;
2. в каждый следующий пул дорожного знака давала бы только те фото, про которые на всех предыдущих знаках сказали "нет".

Это возможно просадило бы качество и не дало бы запускать пулы параллельно для экономии реального времени, но стоимость одного фото ориентировочно была бы равна: ((0.14 + 0.035 + 0.07 + 0.105) + (0.07 \* 4)) / 250 = 0.63 / 250 = **0.252 цента**, что всего в 2.1 раза больше, чем у мультиклассовой классификации.

Насчет **скорости**. На разных дорожных знаках в бинарной классификации она разная. Это может быть связано с тем, что некоторые знаки различать проще остальных. Например, минимальное время выполнения страницы у ярко-желтого Пешеходного перехода, который цепляет глаз. На втором месте Главная дорога, тоже ярко-желтая. Затем Остановка запрещена - красно-синий знак, часто на виду. И замыкают Остановка общественного транспорта и Уступи дорогу.  
В среднем знаки дают время на фото: mean(55, 77, 59, 57, 96) / 44 = 68.8 / 44 = **1.56 сек** - а вот это уже быстрее, чем мультиклассовая классификация.